### PR TITLE
*: extract types.Cluster from etcdserver.Cluster

### PR DESCRIPTION
**Pull request by yichengq**
The PR extracts types.Cluster from etcdserver.Cluster. types.Cluster
is used for flag parsing and etcdserver config.

There is no need to expose etcdserver.Cluster public, which contains
lots of etcdserver internal details and methods. This is the first step
for it.

### CODE REVIEW

File: pkg/types/cluster.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+package types
+
+import (
+	"fmt"
+	"net/url"
+	"sort"
+	"strings"
+)
+
+type Cluster map[string]URLs
```
Comment (xiang90) [inline pkg/types/cluster.go]: i would prefer to make this a struct. we might add more fields into cluster.

File: pkg/types/cluster.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+// NewCluster returns a Cluster instantiated from the given cluster token
+// and cluster string, by parsing members from a set of discovery-formatted
+// names-to-IPs, like:
+// mach0=http://1.1.1.1,mach0=http://2.2.2.2,mach1=http://3.3.3.3,mach2=http://4.4.4.4
+func NewCluster(s string) (Cluster, error) {
+	cl := make(Cluster)
```
Comment (xiang90) [inline pkg/types/cluster.go]: make(Cluster) is very unclear. make should only use to make a map. And you type-define a map as cluster to hide the actual type. and here you leak it again. As a reader, I have to go back to see why you can make a Cluster.

File: pkg/types/cluster.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	v, err := url.ParseQuery(strings.Replace(s, ",", "&", -1))
+	if err != nil {
+		return nil, err
+	}
+	for name, urls := range v {
+		if len(urls) == 0 || urls[0] == "" {
+			return nil, fmt.Errorf("Empty URL given for %q", name)
```
Comment (xiang90) [inline pkg/types/cluster.go]: empty URL for %q. Error strings should not be capitalized

File: pkg/types/cluster.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	for name, peer := range c {
+		for _, url := range peer {
+			kvs = append(kvs, fmt.Sprintf("%s=%s", name, url.String()))
+		}
+	}
+	sort.Strings(kvs)
+	return strings.Join(kvs, ",")
+}
+
+func (c Cluster) Set(name string, urls URLs) { c[name] = urls }
```
Comment (xiang90) [inline pkg/types/cluster.go]: Add. not set. here you treat Cluster as a map again.

File: pkg/types/cluster.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+// Validate ensures that there is no identical urls in the cluster peer list
+func (c Cluster) Validate() error {
+	urlMap := make(map[string]bool)
+	for _, peer := range c {
+		for _, url := range peer {
+			u := url.String()
+			if urlMap[u] {
+				return fmt.Errorf("duplicate url %v in cluster config", url)
```
Comment (xiang90) [inline pkg/types/cluster.go]: what is a cluster config? URL or url? (see line 38)

`duplicate URL %q found in cluster`

File: etcdserver/raft.go [hunk 6]
```diff
@@ -254,18 +255,19 @@ func restartNode(cfg *ServerConfig, snapshot *raftpb.Snapshot) (types.ID, raft.N
 		MaxSizePerMsg:   maxSizePerMsg,
 		MaxInflightMsgs: maxInflightMsgs,
 	}
-	n := raft.RestartNode(c)
+	n = raft.RestartNode(c)
 	raftStatus = n.Status
-	return id, n, s, w
+	return
```
Comment (xiang90) [inline etcdserver/raft.go]: why??? we should almost never use naked return.

File: etcdserver/raft.go [hunk 7]
```diff
@@ -304,9 +306,9 @@ func restartAsStandaloneNode(cfg *ServerConfig, snapshot *raftpb.Snapshot) (type
...
```
Comment (xiang90) [inline etcdserver/raft.go]: see above.

File: etcdserver/raft.go (hunk 6 above)
Comment (xiang90) [inline etcdserver/raft.go]:
```
Naked returns are okay if the function is a handful of lines. Once it's a medium-sized function, be explicit with your return values. Corollary: it's not worth it to name result parameters just because it enables you to use naked returns. Clarity of docs is always more important than saving a line or two in your function.

Finally, in some cases you need to name a result parameter in order to change it in a deferred closure. That is always okay.
```

**Comment by yichengq** [comment]
The main change is to change `Cluster *Cluster` to `InitialCluster types.NameURLPairs` in etcdserver.Config, which decouples how it is set from string and how it is used internally. The benefits of it:

- Cluster struct has many additional info, including cluster ID and member ID, which doesn't need to be set and accessed when building Config.
- There is no need for Cluster to be printed as `name=http://urls` anymore. The new Cluster.String is like this: `{ClusterID:e9c7614f68f35fb2 Members:[&{ID:a8266ecf031671f3 RaftAttributes:{PeerURLs:[http://localhost:7001]} Attributes:{Name:infra1 ClientURLs:[]}} &{ID:6e3bd23ae5f1eae0 RaftAttributes:{PeerURLs:[http://localhost:7002]} Attributes:{Name:infra2 ClientURLs:[]}} &{ID:924e2e83e93f2560 RaftAttributes:{PeerURLs:[http://localhost:7003]} Attributes:{Name:infra3 ClientURLs:[]}}] RemovedMemberIDs:[]}`, which prints out more status.

This also helps make Cluster private because config.Cluster is no longer modified and passed around in etcdserver package.

The code change includes two parts:

- use types.NameURLPairs instead of *etcdserver.Cluster when building Config.
- create *Cluster variable in NewServer instead of modifying config.Cluster. This also helps make Cluster struct private in the future, which cannot be exposed in Config struct then.

The PR doesn't have behavior change.

File: pkg/types/nameurlpairs.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+import (
+	"fmt"
+	"net/url"
+	"sort"
+	"strings"
+)
+
+type NameURLPairs map[string]URLs
+
+// ParseInitialCluster returns a NameURLPairs instantiated from the given string,
```
Comment (xiang90) [inline pkg/types/nameurlpairs.go]: this is so wired. Why does types pkg need to know initial cluster?

File: pkg/types/nameurlpairs.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+package types
+
+import (
+	"fmt"
+	"net/url"
+	"sort"
+	"strings"
+)
+
+type NameURLPairs map[string]URLs
```
Comment (xiang90) [inline pkg/types/nameurlpairs.go]: This seems like a URLsMap

File: pkg/types/nameurlpairs.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+// ParseInitialCluster returns a NameURLPairs instantiated from the given string,
+// which consists of discovery-formatted names-to-URLs, like:
+// mach0=http://1.1.1.1,mach0=http://2.2.2.2,mach1=http://3.3.3.3,mach2=http://4.4.4.4
+func ParseInitialCluster(s string) (NameURLPairs, error) {
```
Comment (xiang90) [inline pkg/types/nameurlpairs.go]: This should be NewURLsMap

File: pkg/types/nameurlpairs.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+		for _, url := range urls {
+			pairs = append(pairs, fmt.Sprintf("%s=%s", name, url.String()))
+		}
+	}
+	sort.Strings(pairs)
+	return strings.Join(pairs, ",")
+}
+
+// Add adds the name-to-URLs pair.
+func (c NameURLPairs) Add(name string, urls URLs) { c[name] = urls }
```
Comment (xiang90) [inline pkg/types/nameurlpairs.go]: This is strictly a set not a add. you are not adding anything into the existing one, but simply replace it.

File: pkg/types/nameurlpairs.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+		for _, u := range us {
+			urls = append(urls, u.String())
+		}
+	}
+	sort.Strings(urls)
+	return urls
+}
+
+// Validate ensures that there is no identical urls in the url list
+func (c NameURLPairs) Validate() error {
```
Comment (xiang90) [inline pkg/types/nameurlpairs.go]: this should not be put in types in my opinion. 

You can define a func as `CheckDuplicateURL(um URLsMap) bool` near where you use it.

**Comment by yichengq** [comment]
all comments addressed.

File: etcdmain/etcd.go [hunk 13]
```diff
@@ -333,35 +335,28 @@ func startProxy(cfg *config) error {
 	return nil
 }
 
-// setupCluster sets up an initial cluster definition for bootstrap or discovery.
-func setupCluster(cfg *config) (*etcdserver.Cluster, error) {
-	var cls *etcdserver.Cluster
-	var err error
+// setupPeerURLsMapAndToken sets up an initial peer URLsMap and cluster token for bootstrap or discovery.
+func setupPeerURLsMapAndToken(cfg *config) (urlsmap types.URLsMap, token string, err error) {
```
Comment (xiang90) [inline etcdmain/etcd.go]: this func is not very clear. it does dnssrv discovery but delay to do the discovery service thing. but this is not the problem of your pull request though.

File: etcdmain/etcd.go [hunk 14]
```diff
@@ -333,35 +335,28 @@ func startProxy(cfg *config) error {
 	return nil
 }
 
-// setupCluster sets up an initial cluster definition for bootstrap or discovery.
-func setupCluster(cfg *config) (*etcdserver.Cluster, error) {
-	var cls *etcdserver.Cluster
-	var err error
+// setupPeerURLsMapAndToken sets up an initial peer URLsMap and cluster token for bootstrap or discovery.
```
Comment (xiang90) [inline etcdmain/etcd.go]: can we change the name to `getPeerURLs...`. It does not set up anything but simply return them now. (yea. it used to set up a cluster struct)

File: etcdserver/config.go [hunk 15]
```diff
@@ -52,10 +52,10 @@ func (c *ServerConfig) VerifyBootstrap() error {
 	if err := c.verifyLocalMember(true); err != nil {
 		return err
 	}
-	if err := c.Cluster.Validate(); err != nil {
-		return err
+	if checkDuplicateURL(c.InitialPeerURLsMap) {
+		return fmt.Errorf("initial cluster has duplicate url")
```
Comment (xiang90) [inline etcdserver/config.go]: initial cluster %s has duplicate url.

File: etcdserver/config.go [hunk 16]
```diff
@@ -70,8 +70,8 @@ func (c *ServerConfig) VerifyJoinExisting() error {
 	if err := c.verifyLocalMember(false); err != nil {
 		return err
 	}
-	if err := c.Cluster.Validate(); err != nil {
-		return err
+	if checkDuplicateURL(c.InitialPeerURLsMap) {
+		return fmt.Errorf("initial cluster has duplicate url")
```
Comment (xiang90) [inline etcdserver/config.go]: same here. we should print out the initial cluster for clarity.

**Comment by xiang90** [comment]
@yichengq Thanks for refactoring this. This pr looks good to me.

File: etcdmain/etcd.go (hunk 13 above)
Comment (yichengq) [inline etcdmain/etcd.go]: agree. i have thoughts about moving this function into etcdserver pkg and let etcdserver do nothing about config parsing.

File: etcdmain/etcd.go (hunk 14 above)
Comment (yichengq) [inline etcdmain/etcd.go]: k

**Comment by xiang90** [comment]
@barakmich Can you take a look?

File: etcdmain/etcd.go [hunk 17]
```diff
@@ -169,21 +169,22 @@ func startEtcd(cfg *config) (<-chan struct{}, error) {
...
-		DiscoveryURL:    cfg.durl,
-		DiscoveryProxy:  cfg.dproxy,
-		NewCluster:      cfg.isNewCluster(),
-		ForceNewCluster: cfg.forceNewCluster,
-		Transport:       pt,
-		TickMs:          cfg.TickMs,
-		ElectionTicks:   cfg.electionTicks(),
+		Name:                cfg.name,
+		ClientURLs:          cfg.acurls,
+		PeerURLs:            cfg.apurls,
```
Comment (barakmich) [inline etcdmain/etcd.go]: nit: Do we still need PeerURLs for ServerConfig, then? Or does InitialPeerURLsMap solve all the cases? And if the latter is true, then we could probably get rid of PeerURLs and change the name of `InitialPeerURLsMap` to `PeerURLsMap`

But only if we can safely remove it.

**Comment by barakmich** [comment]
LGTM if you'd like to merge, one question.

File: etcdmain/etcd.go (hunk 17 above)
Comment (xiang90) [inline etcdmain/etcd.go]: I think we probably can remove this. But I suggest to do it in another pr.

Comment (yichengq) [inline etcdmain/etcd.go]: @barakmich @xiang90 Currently we do sanity checking inside etcdserver package, including whether (cfg.Name, cfg.PeerURLs) appears in cfg.InitialPeerURLsMap. That's why it is here. Unsure about how to make it better.
